[14.0] shopfloor: cluster picking add option first scan location or pack

TDu opened the pull request:
ref.: rau-148

TDu committed:
shopfloor: cp add option first scan location or pack

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -416,7 +417,7 @@ def unassign(self, picking_batch_id):
         return self._response_for_start()
 
-    def scan_line(self, picking_batch_id, move_line_id, barcode):
+    def scan_line(self, picking_batch_id, move_line_id, barcode, sublocation_id=None):
```
```suggestion
    def scan_line(self, picking_batch_id, move_line_id, barcode, package=None):
```

TDu replied:
After discussion we do not need to pass the package between the frontend and backend.
But I still believe we need to do this for the sublocation, example :

> The user scan a location containing multiple products. He will first scan the location but the current implementation will ask him to scan the product. And when the scan line or packaging first option is enabled the backend needs to remember that a location was scanned previously

.

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -456,17 +463,21 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
-        product = search.product_from_scan(barcode)
-        if product and move_line.product_id == product:
-            return self._scan_line_by_product(picking, move_line, product)
+        product = self.env["product.product"]
+        if not self.work.menu.scan_location_or_pack_first or sublocation:
+            product = search.product_from_scan(barcode)
+            if product and move_line.product_id == product:
+                return self._scan_line_by_product(picking, move_line, product)
```
```suggestion
        product = search.product_from_scan(barcode)
        if product and move_line.product_id == product:
            if self.work.menu.scan_location_or_pack_first:
              if move_line.package_id == package or False:
                 return self._scan_line_by_product(picking, move_line, product)
            else:
               return self._scan_line_by_product(picking, move_line, product)
```

jbaudoux reviewed (changes requested):
You need to give the package in the context

jbaudoux commented on shopfloor/services/cluster_picking.py:
```diff
@@ -488,7 +507,31 @@ def _get_prefill_qty(self, move_line, qty=0):
+            )
+        return None
+
+    def _scan_line_by_package(self, picking, move_line, package, batch, sublocation):
```
I would drop this section so that we always jump to next screen when the right package is scanned. The next screen will take care of scanning the product/lot

```
        if packaging.package_has_several_products(package):
            message = self.msg_store.several_products_in_package(package)
        elif packaging.package_has_several_lots(package):
            message = self.msg_store.several_lots_in_package(package)
        if message:
            return self._pick_next_line(batch, message=message)
```

TDu replied:
Removed

TDu committed:
shopfloor: cp scan_line remove message

Removed because we always want to jump to the next screen when the right
package is scanned. The scanning of the product or lot will be done
on the next screen.
Also the use of `pick_next_line` when there is an error message does not
make sense.

simahawk commented:
@TDu can you rebase?

jbaudoux commented:
/ocabot merge minor
